examples/examples/jsonrpsee_as_service.rs

```diff
@@ -70,26 +75,97 @@ struct AuthorizationMiddleware<S> {
 	transport_label: &'static str,
 }
 
-impl<'a, S> RpcServiceT<'a> for AuthorizationMiddleware<S>
+impl<S> AuthorizationMiddleware<S> {
+	/// Authorize the request by checking the `Authorization` header.
+	///
+	///
+	/// In this example for simplicity, the authorization value is not checked
+	// and used because it's just a toy example.
+	fn auth_method_call(&self, req: &Request<'_>) -> bool {
+		if req.method_name() == "trusted_call" {
+			let Some(Ok(_)) = self.headers.get(AUTHORIZATION).map(|auth| auth.to_str()) else { return false };
+		}
+
+		true
+	}
+
+	/// Authorize the notification by checking the `Authorization` header.
+	///
+	/// Because notifications are not expected to return a response, we
+	/// return a `MethodResponse` by injecting an error into the extensions
+	/// which could be read by other middleware or the server.
+	fn auth_notif(&self, notif: &Notification<'_>) -> bool {
+		if notif.method_name() == "trusted_call" {
+			let Some(Ok(_)) = self.headers.get(AUTHORIZATION).map(|auth| auth.to_str()) else { return false };
+		}
+
+		true
+	}
+}
+
+impl<S> RpcServiceT for AuthorizationMiddleware<S>
 where
-	S: Send + Clone + Sync + RpcServiceT<'a>,
+	S: Send + Clone + Sync + RpcServiceT<Response = MethodResponse, Error = Infallible> + 'static,
+	S::Error: Into<BoxError> + Send + 'static,
+	S::Response: Send,
 {
-	type Future = ResponseFuture<S::Future>;
+	type Error = S::Error;
+	type Response = S::Response;
```

Comment:
This could be a followup thing to try out but: I wonder whether it's worth now having a separate response type for each call, so that there is no enum?

eg

``rust
type CallResponse = ...
type BatchResponse = ...
type NotificationResponse = ...
```

Reply:
(though maybe this prevents the middleware from doing certain things, ie returning a MethodResponse::Error in response to a call or whatever?)

Reply:
Let's file an issue for this, it's not trivial to break up the method response into different types in the server